pkg/kubelet/kubelet_pods.go

```diff
@@ -915,6 +915,12 @@ func countRunningContainerStatus(status v1.PodStatus) int {
 	return runningContainers
 }
 
+// PodCouldHaveRunningContainers returns true if the pod with the given UID could still have running
+// containers. This returns false if the pod has not yet been started or the pod is unknown.
+func (kl *Kubelet) PodCouldHaveRunningContainers(pod *v1.Pod) bool {
+	return kl.podWorkers.CouldHaveRunningContainers(pod.UID)
```

Comment:
Do we also need to check the status to ensure no containers are running? i.e. the check here - https://github.com/kubernetes/kubernetes/blob/f4df7d0eb27b51a0f47a49616b8b32e0bf9c18a7/pkg/kubelet/kubelet_pods.go#L926-L931

I'm not sure if this check is redundant or not, but at least in `PodResourcesAreReclaimed` it checks both `podWorkers.CouldHaveRunningContainers` as well as the running containers from the status...

Reply:
The check in PodResourcesReclaimed isn't checking for running containers - it's checking to see that we have received all the container statuses from the status loop before we delete the pod (we're trying to make sure we propagate all the expected statuses).  TerminatePod called within syncTerminatedPod guarantees that those statuses are flushed - the wait guarantees we write them to the status before we finally delete the pod (which is the last possible time to write the container status).

PodCouldHaveRunning containers is weaker - we don't need to be able to write all the container status before we write the phase, as long as it is written later.  So we need to wait until those statuses propagate.